File: JCleanArchitectureApp/appdagger/src/main/java/com/woowrale/jcleanarchitecture/di/factory/UseCaseFactory.java
package com.woowrale.jcleanarchitecture.di.factory;

import com.woowrale.data.repository.local.LocalRepository;
import com.woowrale.data.repository.remote.RemoteRepository;
import com.woowrale.usecase.base.BaseUseCase;
import com.woowrale.usecase.threads.JobScheduler;
import com.woowrale.usecase.threads.UIScheduler;
import com.woowrale.usecase.usescases.GetContactAllUseCase;
import com.woowrale.usecase.usescases.GetContactsUseCase;

import javax.inject.Inject;

public class UseCaseFactory {

    private LocalRepository localRepository;
    private RemoteRepository remoteRepository;
    private UIScheduler uiScheduler;
    private JobScheduler jobScheduler;

    @Inject
    public UseCaseFactory(RemoteRepository remoteRepository,
                          LocalRepository localRepository,
                          UIScheduler uiScheduler,
                          JobScheduler jobScheduler){

        this.remoteRepository = remoteRepository;
        this.localRepository = localRepository;
        this.uiScheduler = uiScheduler;
        this.jobScheduler = jobScheduler;
    }

    public BaseUseCase getRemoteContacts(){
        return new GetContactsUseCase(remoteRepository, uiScheduler, jobScheduler);
    }

    public BaseUseCase getLocalContacts(){
        return new GetContactAllUseCase(localRepository, uiScheduler, jobScheduler);
    }
}

File: KCleanArchitectureApp/appdagger/src/main/java/com/woowrale/kcleanarchitecture/di/factory/UseCaseFactory.kt
package com.woowrale.kcleanarchitecture.di.factory

import com.woowrale.data.repository.local.LocalRepository
import com.woowrale.data.repository.remote.RemoteRepository
import com.woowrale.usecase.threads.JobScheduler
import com.woowrale.usecase.threads.UIScheduler
import com.woowrale.usecase.usecases.GetContactAllUseCase
import com.woowrale.usecase.usecases.GetContactsUseCase
import javax.inject.Inject

class UseCaseFactory @Inject constructor(
    private val remoteRepository: RemoteRepository,
    private val localRepository: LocalRepository,
    private val uiScheduler: UIScheduler,
    private val jobScheduler: JobScheduler
) {

    fun getRemoteContacts(): GetContactsUseCase {
       return GetContactsUseCase(remoteRepository, uiScheduler, jobScheduler)
    }

    fun getLocalContacts(): GetContactAllUseCase{
        return  GetContactAllUseCase(localRepository, uiScheduler, jobScheduler)
    }
}